Cleaned up version of earlier PR #48601.

PeterSolMS opened the pull request:
Port of ConserveMemory setting we implemented on .NET Framework.

PeterSolMS committed:
Cleaned up version of earlier PR #48601.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -18285,7 +18298,38 @@ int gc_heap::joined_generation_to_condemn (BOOL should_evaluate_elevation,
+        }
+    }
+
+    if ((conserve_mem_setting != 0) && (n >= max_generation))
```
n cannot be > max_generation... it can only be at most max_generation.

PeterSolMS replied:
I know - I changed the >= to ==.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23360,6 +23404,21 @@ size_t gc_heap::get_total_gen_estimated_reclaim (int gen_number)
+    for (int hn = 0; hn < gc_heap::n_heaps; hn++)
+    {
+        gc_heap* hp = gc_heap::g_heaps[hn];
+        size += hp->generation_sizes (hp->generation_of (gen_number));
```
this should use `generation_size`, not `generation_sizes`

PeterSolMS replied:
Yup, thanks for catching!

PeterSolMS committed:
Address code review feedback.

PeterSolMS committed:
Take out changes in dt_high_frag_p - probably no longer necessary.

Remove changes for the heap_hard_limit case - we'll do those later.

Fixed another place where I had called generation_sizes instead of generation_size.

PeterSolMS committed:
Fixed parameter for call to generation_size.

PeterSolMS committed:
Update comment to the more explicit version in .NET Framework.

PeterSolMS committed:
Added parentheses around an expression to address code review feedback.

Maoni0 reviewed (approved):
LGTM!

PeterSolMS commented:
/backport to release/6.0
